src/atoms/affine/VcatAtom.jl

```diff
@@ -0,0 +1,61 @@
+# Copyright (c) 2014: Madeleine Udell and contributors
+#
+# Use of this source code is governed by a BSD-style license that can be found
+# in the LICENSE file or at https://opensource.org/license/bsd-2-clause
+
+mutable struct VcatAtom <: AbstractExpr
+    children::Tuple
+    size::Tuple{Int,Int}
+
+    function VcatAtom(args...)
+        args = convert.(AbstractExpr, args)
+        num_rows, num_cols = 0, args[1].size[2]
+        for arg in args
+            if arg.size[2] != num_cols
+                msg = "[VcatAtom] cannot stack expressions of incompatible size. Got $(arg.size[2]) expected $num_cols."
+                throw(DimensionMismatch(msg))
+            end
+            num_rows += arg.size[1]
+        end
+        return new(args, (num_rows, num_cols))
+    end
+end
+
+head(io::IO, ::VcatAtom) = print(io, "vcat")
+
+Base.sign(x::VcatAtom) = sum(map(sign, x.children))
+
+monotonicity(x::VcatAtom) = ntuple(_ -> Nondecreasing(), length(x.children))
+
+curvature(::VcatAtom) = ConstVexity()
+
+evaluate(x::VcatAtom) = reduce(vcat, map(evaluate, x.children))
+
+function new_conic_form!(context::Context{T}, x::VcatAtom) where {T}
+    # Converting a VcatAtom to conic form is non-trivial. Consider two matrices:
+    #   x = [1 3; 2 4]
+    #   y = [5 7; 6 8]
+    # with VcatAtom(x, y). The desired outcome is [1, 2, 5, 6, 3, 4, 7, 8].
+    # If we naively convert the children to conic form and then vcat, we will
+    # get:
+    #   vcat([1, 2, 3, 4], [5, 6, 7, 8]) = [1, 2, 3, 4, 5, 6, 7, 8]
+    # which is not what we are after. We need to first transpose each child to
+    # get:
+    #   x^T, y^T = [1 2; 3 4], [5 6; 7 8])
+    # then hcat them to get:
+    #   hcat(x^T, y^T) = [1 2 5 6; 3 4 7 8]
+    # then transpose this to get:
+    #   hcat(x^T, y^T)^T = [1 3; 2 4; 5 7; 6 8]
+    # so our final connic form produces the desired
+    #   [1, 2, 5, 6, 3, 4, 7, 8]
+    return conic_form!(context, transpose(reduce(hcat, transpose.(x.children))))
```

Comment:
This part is not nice, but I don't know an alternative.

Reply:
One option is to manually build the operator we want to do: https://github.com/jump-dev/Convex.jl/blob/0bd470bafbaecdd165a71708a6bf745ebe9b550b/src/reformulations/partialtranspose.jl#L44

Reply:
actually, I think that `permutedims_matrix` function already does what we need:
```julia
julia> x = [1 3; 2 4]
2×2 Matrix{Int64}:
 1  3
 2  4

julia> y = [5 7; 6 8]
2×2 Matrix{Int64}:
 5  7
 6  8

julia> M = permutedims_matrix((size(x, 1),size(x,2),size(y,1)), (1,3,2))
8×8 SparseMatrixCSC{Bool, Int64} with 8 stored entries:
 1  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  1  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  1  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  1  ⋅  ⋅
 ⋅  ⋅  1  ⋅  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  1  ⋅  ⋅  ⋅  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  1  ⋅
 ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  1

julia> reshape(M * vcat(vec(x),vec(y)), size(x,1) + size(y,1), size(x,2))
4×2 Matrix{Int64}:
 1  3
 2  4
 5  7
 6  8
```

Reply:
I've kept as-is for now. It wasn't obvious how to generalize this to the n-ary case, and what we currently have works.

Reply:
We can try this refactoring in a separate PR